ESPS/general/src/addfea/addfea.1
.\" Copyright (c) 1987, 1990 Entropic Speech, Inc.; All rights reserved
.\" @(#)addfea.1	1.11 1.11 ESI
.TH ADDFEA 1\-ESPS 4/2/97
.ds ]W "\fI\s+4\ze\h'0.05'e\s-4\v'-0.4m'\fP\(*p\v'0.4m'\ Entropic Speech, Inc.
.SH "NAME"
addfea - adds a new FEA file field based on ASCII data 
.SH "SYNOPSIS"
.B addfea
[
.BI \-P
.I params
] [
.BI \-f
.I field_name
] [
.BI \-t
.I field_type
] [
.BI \-s
.I field_size
] [
.BI \-T
.I FEA_subtype
] [
.BI \-x " debug_level"
] [
.BI \-c " comment"
] [
.BI \-C " comment_file"
]
.I "ascfile feafile"
[
.I "feafile.out"
]
.SH "DESCRIPTION"
.PP
\fIAddfea\fR creates a new FEA file field
and fills it with data taken from the ASCII file 
.I ascfile.  
If \fIfeafile.out\fP is not supplied, then the new field 
is added to each record of \fIfeafile\fP; if 
\fIfeafile\fP does not exist, a new file is created with one field per
record.  If \fIfeafile.out\fP is supplied, then the records from 
\fIfeafile\fP are copied to \fIfeafile.out\fP with the new field 
being added to each record.  In this case, \fIfeafile\fP must exist.  
.PP
If \fIascfile\fP is "\-", standard input is used for the input ASCII file.  If
\fIfeafile\fP is "\-", standard input is used for the input FEA file.  (It
is an error for both \fIascfile\fP and \fIfeafile\fP to be "\-".)  If
\fIfeafile.out\fP is "\-", standard output is used.  If \fIfeafile\fP
exists and is not an ESPS FEA file, \fIaddfea\fP prints an error message
and exits.
.PP
\fIAscfile\fP is an ASCII file containing values separated by white
space or new lines.  \fIAddfea\fP takes \fIfield_size\fP values at a
time from \fIascfile\fP, converts them to type \fIfield_type\fP, and
stores them in a field named \fIfield_name\fP within sequential
records of \fIfeafile\fP.  For complex types, each complex value is
represented by a two ASCII inputs \- real part followed by imaginary
part.  \fIAddfea\fP keeps processing the input until it is exhausted
or until all records in \fIfeafile\fP (if it exists) have been
updated.  If the input is exhausted before all records have been
updated, the \fIfea_field\fP in remaining records is set to zero and a
warning is issued.  If additional input remains after all records have
been updated, the additional input is discarded and a warning is
issued.  If \fIascfile\fP is empty, then the new FEA field is created
and set uniformly to zero.
.PP
If the field name given for the \fB-f\fR option is \fITag\fR, the built
in tag value of the record is modified, rather than a feature field
named \fITag\fR.   This is a special case, to allow setting the built in
tag value on each feature record.   See \fIfea(5\-ESPS)\fR for details
on the tag field.   
In this case, the other field related options are ignored (size and
type).
.PP
If the new field is complex data type, then two values from the ASCII
file are used for each data item.  The real value is given first, and
then the imaginary value.  So a complex field of size 2 would
require 4 values.  If an odd number of ASCII values is given, then the
imaginary part is filled with zero.
.PP
For record-keeping purposes,
.I addfea
stores an ASCII comment in the header of of the output file.  The comment
should describe the origins of the ASCII data in \fIascfile\fP.  If a
comment is not supplied in the parameter file or by means of the \fB\-c\fR
or \fB\-C\fR options, the user is prompted for one.  If \fIaddfea\fP
is called on a pipe with \fIascfile\fP or \fIfeafile\fP as 
standard input, the comment must be supplied directly as the user
cannot be prompted.  
.PP
Note that \fIpplain\fP (1\-\s-1ESPS\s+1) and \fIselect\fP (1\-\s-1ESPS\s+1) 
(in EVAL mode) can be used to create ASCII inputs for \fIaddfea\fP.  
.PP
Note that the difference between CHAR and BYTE is that CHAR is for
character data (either a single character or a character string) and
BYTE is for byte signed integer data.
Also,
CHAR data is handled a bit awkwardly. A new-line is considered a deliminator
for a string of characters,
but it also is retained in the string. So, for example, if you
have four lines of ASCII data
.sp
.nf
This
is
a
test

.fi
and you want to have each output record contain a single word,
you must specify a size of 5 or larger.
That is, you must allow space for the new-line character.
.SH OPTIONS
.PP
The following options are supported:
.TP
.BI \-P " param" " \fR[params]\fP"
Specifies the name of the parameter file.
.TP
.BI \-f " field_name" " "
Specifies the name of the new FEA field.  If a field with this name
already exists in \fIfeafile\fP, \fIaddfea\fR exits with an error message.
If this option is used, it overides the parameter file value.
The field name \fITag\fR is handled as a special case as described above.
.TP
.BI \-t " field_type" " \fR[FLOAT]\fP"
Specifies the type of the new field.  Allowable types are the
following: DOUBLE, FLOAT, LONG, SHORT, CHAR, BYTE, DOUBLE_CPLX,
FLOAT_CPLX, LONG_CPLX, SHORT_CPLX, CHAR_CPLX, and BYTE_CPLX (case
doesn't matter).  If this option is used, it overides the parameter
file value.
.TP
.BI \-s " field_size" " \fR[1]\fP"
Specifies the size of the new field.  If this option is used, it overides
the parameter file value.
.TP
.BI \-T " FEA_subtype " 
If this option is given, the feature file sub-type code is filled in
with the given code.  The symbolic name of the sub-type code must be
given, for example FEA_ANA or FEA_SD.  For a complete list of defined
sub-type codes, see FEA(5\-ESPS).  Use of this option doesn't ensure
that the is of the correct format for the defined subtype.  The user
should use this option only in cases where he or she knows that a file
with the correct fields is being produced.  See the section 5 man
pages for the defined feature file types.
.TP
.BI \-x " debug_level \fR[0]\fP"
If 
.I debug_level
is positive,
.I testsd
prints debugging messages and other information on the standard error
output.  The messages proliferate as the  
.I debug_level
increases.  If
.I debug_level
is 0, no messages are printed.  Only levels 0 and 1 are supported.   
.TP
.BI \-c " comment"
Specifies that the ASCII string 
.I comment
be added as a comment in the header of the output file.  If this option is
used, it overides the parameter file value.
.TP
.BI \-C " comment_file"
Specifies that the contents of the file
.I comment_file
be added as a comment in the header of the output file.  If this option is
used, it overides the parameter file value.
.SH "ESPS PARAMETERS"
.PP
The parameter file does not have to be present if all of the parameters are
specified by means of command-line options.  The following parameters are
read, if present, from the parameter file:
.IP
.I "field_name \- string"
.IP
This is the name of the field to add to \fIfeafile\fP.
.IP
.I "field_type \- string"
.IP
This is the type of the field to add to \fIfeafile\fP.  Allowable
values are DOUBLE, FLOAT, LONG, SHORT, CHAR, BYTE, DOUBLE_CPLX,
FLOAT_CPLX, LONG_CPLX, SHORT_CPLX, CHAR_CPLX, and BYTE_CPLX (case
doesn't matter).
.IP
.I "field_size \- integer"
.IP
This specifies the size of \fIfield_name\fP \- i.e., the size of the
field added to \fIfeafile\fP.
.IP
.I "comment \- string"
.IP
This provides an ASCII comment that is added to the comment field in the
output file.  The comment should describe the origins of the data in
\fIascfile\fP.  If this parameter is not present in the parameter file and
is not specified by the \fB\-c\fP or \fB\-C\fP options, then the user is
prompted for a comment provided that \fIaddfea\fP is not called on a pipe
with \fIascfile\fP or \fIfeafile\fP as standard input.
.PP
Remember that command line option values override 
parameter file values.
.SH "ESPS COMMON"
.PP
ESPS Common is not read or written by \fIaddfea\fP.  
.SH ESPS HEADERS
.PP
All the old header values are copied to the new header and the input FEA
file becomes a source file within the recursive header.
\fIaddgen\fP (1\-\s-1ESPS\s+1) can be used to add the \fIstart_time\fP
and \fIrecord_freq\fP generics used by \fIwaves+\fP,
if they don't already exist.
.SH "FUTURE CHANGES"
.PP
Add CODED type to the list of possible field types.
.PP
Improve the handling of CHAR data. 
Make each line correspond to a data record.
Allow use of either \fIfield_size\fP characters or the new-line character
as the end-of-data deliminator for each line of input.
If more than \fIfield_size\fP characters exist on a line,
a warning is issued and only \fIfield_size\fP characters are used.
If less than \fIfield_size\fP characters occur before a new-line character,
the string is null padded.
.SH "SEE ALSO"
.PP
.nf
\fIaddgen\fP (1\-\s-1ESPS\s+1), \fIcomment\fP(1\-\s-1ESPS\s+1), \fIpsps\fP(1\-\s-1ESPS\s+1), 
\fIpplain\fP (1\-\s-1ESPS\s+1), \fIselect\fP(1\-\s-1ESPS\s+1), \fIfea_deriv\fP (1\-\s-1ESPS\s+1)
.fi
.SH "WARNINGS AND DIAGNOSTICS"
.PP
\fIAddfea\fP will exit with an error message if \fIfeafile\fP exists and
already contains a field named \fPfield_name\fP.  
.PP
\fIAddfea\fP will exit with an error message if both \fIascfile\fP 
and \fIfeafile\fP are "\-".  
.PP
\fIAddfea\fP will exit with an error message if, when \fIaddfea\fP is
called on a pipe with \fIascfile\fP as standard input, a comment is not
supplied in the parameter file or via \fB\-c\fP or \fB\-C\fP.  
.SH "BUGS"
.PP
None known.  
.SH "AUTHOR"
.PP
Manual page by John Shore.  



ESPS/general/src/addfea/addfea.c
/*
 * This material contains unpublished, proprietary software of 
 * Entropic Research Laboratory, Inc. Any reproduction, distribution, 
 * or publication of this work must be authorized in writing by Entropic 
 * Research Laboratory, Inc., and must bear the notice: 
 *
 *    "Copyright (c) 1986-1990  Entropic Speech, Inc. 
 *    "Copyright (c) 1990-1991  Entropic Research Laboratory, Inc. 
 *                   All rights reserved"
 *
 * The copyright notice above does not evidence any actual or intended 
 * publication of this source code.     
 *
 * Written by:  Alan Parker
 * Checked by:
 * Revised by: Ken Nelson
 *
 * Brief description: creates new fea file feidl and fills it with data
 *
 */

static char *sccs_id = "@(#)addfea.c	1.20	7/1/93	ESI/ERL";

int     debug_level = 0;


#ifndef lint
#define DATE "7/1/93"
#define VERS "1.20"
#else
#define DATE "none"
#define VERS "none"
#endif

#define SYNTAX USAGE("addfea [-P param] [-f field_name] [-t field_type] [-s field_size] \n\t[-T subtype_code] [-x debug_level] [-c comment] [-C comment_file] ascfile feafile [feafile.out]")
#define COMMENT_LINE_SIZE 512

#include <stdio.h>
#include <esps/esps.h>
#include <esps/fea.h>
#include <esps/unix.h>

extern int  optind;
extern char *optarg;
char    sbuf[256];
FILE * fopen ();
char   *get_cmd_line ();
void copy_fea_rec ();
char  *savestring();
char *add_lf();

#define error(s,f) error2(s," ",f)
void
error2 (s, t, flag)
char   *s,
       *t;
int     flag;
{
    (void) fprintf (stderr, "addfea: %s %s\n", s, t);
    if (flag)
	SYNTAX;
    exit (1);
}

void
get_comment (file, hd)
char   *file;
struct header  *hd;
{
    FILE * strm;
    char   *buffer = malloc ((unsigned) COMMENT_LINE_SIZE);

    spsassert (buffer, "malloc failed!");
    if (strcmp ("<stdin>", file) == 0) {
	fprintf (stderr, "addfea: Please enter comment, ended with blank line:\n");
	strm = stdin;
    }
    else
    if ((strm = fopen (file, "r")) == NULL)
	error2 ("cannot open comment file:", file, 0);

    while (fgets (buffer, COMMENT_LINE_SIZE, strm) != NULL) {
	if (strm == stdin && strlen (buffer) == 1)
	    return;
	add_comment (hd, buffer);
    }
    free (buffer);
}

char *
add_lf(line)
char *line;			/* line to add a line feed to */
{
  char *newline = malloc((unsigned) strlen(line) + 2);
  return(savestring(strcat(strcpy(newline, line),"\n")));
}

main (argc, argv)
int     argc;
char  **argv;
{
    char   *out_file = NULL;
    char   *in_file = NULL;
    char   *asc_file = NULL;
    FILE * asc_strm = NULL;
    FILE * in_strm = NULL;
    FILE * out_strm = NULL;
    char   *subtype = NULL;
    int    subtype_code = 0;
    char   *comment = NULL;
    char   *comment_file = NULL;
    struct header  *in_hd,
                   *out_hd;
    struct fea_data *out_rec,
                   *in_rec;
    int     c,
            new_file = 0,	/* 1 when output file is new */
            f_flag = 0,
            t_flag = 0,
            s_flag = 0,
            update = 0,		/* 1 when updating input file */
	    add_tag=0;
    short   type;
    char   *ptr;
    double  dbuf;
    int     size;
    long   *l_ptr;
    short  *s_ptr;
    float  *f_ptr;
    double *d_ptr;
    char   *c_ptr;
    double_cplx *dc_ptr;
    float_cplx *fc_ptr;
    long_cplx *lc_ptr;
    short_cplx *sc_ptr;
    byte_cplx *bc_ptr;

/* default parameters */

    char   *param_file = NULL;/* default parameter file name */
    char   *field_name = NULL;	/* NO default field_name */
    char   *field_type = "FLOAT";/* default field type */
    long    field_size = 1;	/* default field_size */

    while ((c = getopt (argc, argv, "P:f:t:s:c:C:T:x:")) != EOF) {
	switch (c) {
	    case 'P': 
		param_file = optarg;
		break;
	    case 'f': 
		f_flag = 1;
		field_name = optarg;
		break;
	    case 't': 
		t_flag = 1;
		field_type = optarg;
		break;
	    case 's': 
		s_flag = 1;
		field_size = atoi (optarg);
		break;
	    case 'c': 
		comment = add_lf(optarg);
		break;
	    case 'C': 
		comment_file = optarg;
		break;
	    case 'T':
		subtype = optarg;
		break;
	    case 'x': 
		debug_level = 1;
		break;
	}
    }
    if ((argc - optind) < 2)
	error ("at least two files must be given", 1);

    if (field_name && strcmp(field_name,"Tag") == 0) {
	add_tag = 1;
	t_flag=1;
	field_type = "LONG";
    }

    if (comment != NULL && comment_file != NULL)
	error ("cannot give both -C and -c options", 1);

    if (!(f_flag && t_flag && s_flag && (comment || comment_file))) {
/* need to read parameter file */
	(void) read_params (param_file, SC_NOCOMMON, NULL);
	if (!f_flag && (symtype ("field_name") == ST_STRING))
	    field_name = getsym_s ("field_name");
	if (!t_flag && (symtype ("field_type") == ST_STRING))
	    field_type = getsym_s ("field_type");
	if (!s_flag && (symtype ("field_size") == ST_INT))
	    field_size = getsym_i ("field_size");

	if (!subtype && (symtype ("FEA_subtype") == ST_STRING))
	    subtype = getsym_s ("FEA_subtype");

	if (!(comment || comment_file) && symtype ("comment") ==
		ST_STRING)
	    comment = getsym_s ("comment");
    }



    if ((argc - optind) == 3)
	out_file = argv[optind + 2];

    asc_file = argv[optind];
    in_file = argv[optind + 1];

    if (strcmp (asc_file, in_file) == 0)
	error ("ascfile and feafile cannot be the same", 1);


    if (field_name == NULL)
	error("no \"field_name\" specifed - exiting", 0);

    if (field_type == NULL)
	error ("no \"field_type\" specified - exiting.", 0);

    if (debug_level) {
	fprintf (stderr, "comment: %s\n", comment);
	fprintf (stderr, "comment_file %s\n", comment_file);
	fprintf (stderr, "asc_file: %s\n", asc_file);
	fprintf (stderr, "in_file: %s\n", in_file);
	fprintf (stderr, "out_file: %s\n", out_file);
	fprintf (stderr, "param_file: %s\n", param_file);
	fprintf (stderr, "field_name: %s\n", field_name);
	fprintf (stderr, "field_type: %s\n", field_type);
	fprintf (stderr, "field_size: %d\n", field_size);
    }

    type = lin_search (type_codes, field_type);
    if (type == -1)
	error2 ("unknown type code:", field_type, 0);
    if (type == CODED)
	error ("type CODED not supported yet", 0);

/* 
  I've already checked that asc_file and in_file aren't both the same
*/
    if (strcmp (asc_file, "-") == 0)
	asc_strm = stdin;
    else
    if (!(asc_strm = fopen (asc_file, "r")))
	error2 ("cannot open ASCII data file:", asc_file, 0);

/* 
  If a third filename is given, then take it to be an output file to
  open for writing.  In this case, the second filename must be an
  existing feature file.
  
*/

    if (out_file) {
	new_file = 1;
	out_file = eopen ("addfea", out_file, "w", FT_FEA, NONE,
		&out_hd, &out_strm);

	in_file = eopen ("addfea", in_file, "r", FT_FEA, NONE,
		&in_hd, &in_strm);
	in_rec = allo_fea_rec (in_hd);
    }
    else {
/*
  An output file is not given, so the second name might be either an
  existing file for update, or a non-existing file for output
*/

	if ((in_strm = fopen (in_file, "r")) == NULL) {
	    /* file does not exist */
	    if(add_tag)
		error("cannot create a new file with just a tag", 0);
	    new_file = 1;
	    in_file = eopen ("addfea", in_file, "w", FT_FEA, NONE,
		    &out_hd, &out_strm);
	}
	else {
	    /* file does exist, to be updated */
	    fclose (in_strm);	/* prefer to use eopen */
	    in_file = eopen ("addfea", in_file, "r", FT_FEA, NONE,
		    &in_hd, &in_strm);
	    in_rec = allo_fea_rec (in_hd);
	    update = 1;
	    out_strm = tmpfile ();
	}
    }


    if (new_file & !out_file) {
	out_hd = new_header (FT_FEA);
    }
    else {
	out_hd = copy_header (in_hd);
	add_source_file (out_hd, in_file, in_hd);
    }
    strcpy (out_hd -> common.prog, "addfea");
    strcpy (out_hd -> common.progdate, DATE);
    strcpy (out_hd -> common.vers, VERS);

    if (add_tag) {
	out_hd->common.tag = 1;
    }

    add_comment (out_hd, get_cmd_line (argc, argv));
    sprintf (sbuf, "Field %s added by addfea\n", field_name);
    add_comment (out_hd, sbuf);

    if (comment_file)
	get_comment (comment_file, out_hd);
    else if (comment) 
	add_comment (out_hd, comment);
    else if (strcmp ("-", asc_file) && strcmp ("<stdin>", in_file))
	get_comment ("<stdin>", out_hd);
    else
	error ("you must give a comment on the command line or in the parameter file", 0);

    if (field_size < 1)
	error ("the field size cannot be less than 1", 0);

    if (!add_tag && 
	(add_fea_fld (field_name, field_size, (short) 1, (long *) NULL, type,
		(char **) NULL, out_hd) == -1))
	error2 ("field already in output file:", field_name, 0);

    /*set subtype code if called for */

    if (subtype) {
       if ((subtype_code = lin_search(fea_file_type, subtype)) == -1)
	  Fprintf(stderr,"addfeahd: unknown FEA file subtype %s ignored.\n",
		subtype);
       else
	  out_hd->hd.fea->fea_type = subtype_code;
     }

	

    write_header (out_hd, out_strm);

    out_rec = allo_fea_rec (out_hd);
    size = field_size;
    if (!add_tag) {
    	ptr = get_fea_ptr (out_rec, field_name, out_hd);
    	spsassert (ptr != NULL, "get_fea_ptr returned null");

    	size = field_size;
    	d_ptr = (double *) ptr;
    	f_ptr = (float *) ptr;
    	l_ptr = (long *) ptr;
    	s_ptr = (short *) ptr;
    	c_ptr = ptr;
    	dc_ptr = (double_cplx *) ptr;
    	fc_ptr = (float_cplx *) ptr;
    	lc_ptr = (long_cplx *) ptr;
    	sc_ptr = (short_cplx *) ptr;
    	bc_ptr = (byte_cplx *) ptr;
     }
     else
	l_ptr = &(out_rec->tag);

    while (1) {
      int retval = 0;
	if (type != CHAR) {
	    if ((retval = fscanf (asc_strm, "%lf", &dbuf)) == EOF)
		break;
	    if(retval == 0){
	      Fprintf(stderr, 
"addfea: Data type in file \"%s\" and specified value (%s) differ - exiting.\n",
		      asc_file, field_type);
	      exit(1);
	    }
        }
	else if (type == CHAR && field_size != 1) {
	    if (fgets (ptr, size+1, asc_strm) == NULL)
	    	break;
	}
	else if (type == CHAR && field_size == 1) {
	    if (fscanf (asc_strm, "%c", ptr) == EOF)
		break;
	}
	switch (type) {
	    case DOUBLE: 
		*d_ptr++ = dbuf;
		break;
	    case FLOAT: 
		*f_ptr++ = dbuf;
		break;
	    case LONG: 
		*l_ptr++ = dbuf;
		break;
	    case SHORT: 
		*s_ptr++ = dbuf;
		break;
	    case BYTE:
		*c_ptr++ = dbuf;
		break;
	    case CHAR: 
		size = 1;
		break;
	    case DOUBLE_CPLX:
		dc_ptr->real = dbuf;
	    	if (fscanf (asc_strm, "%lf", &dbuf) == EOF)
			dc_ptr->imag = 0;
		else
			dc_ptr->imag = dbuf;
		dc_ptr++;
		break;
	    case FLOAT_CPLX:
		fc_ptr->real = dbuf;
	    	if (fscanf (asc_strm, "%lf", &dbuf) == EOF)
			fc_ptr->imag = 0;
		else
			fc_ptr->imag = dbuf;
		fc_ptr++;
		break;
	    case LONG_CPLX:
		lc_ptr->real = dbuf;
	    	if (fscanf (asc_strm, "%lf", &dbuf) == EOF)
			lc_ptr->imag = 0;
		else
			lc_ptr->imag = dbuf;
		lc_ptr++;
		break;
	    case SHORT_CPLX:
		sc_ptr->real = dbuf;
	    	if (fscanf (asc_strm, "%lf", &dbuf) == EOF)
			sc_ptr->imag = 0;
		else
			sc_ptr->imag = dbuf;
		sc_ptr++;
		break;
	    case BYTE_CPLX:
		bc_ptr->real = dbuf;
	    	if (fscanf (asc_strm, "%lf", &dbuf) == EOF)
			bc_ptr->imag = 0;
		else
			bc_ptr->imag = dbuf;
		bc_ptr++;
		break;
	    default: 
		error ("internal error, should not be here #1", 0);
	      }

	if (size-- == 1) {
	    if (new_file && !out_file)
		put_fea_rec (out_rec, out_hd, out_strm);
	    else
	    if (get_fea_rec (in_rec, in_hd, in_strm) != EOF) {
		copy_fea_rec (in_rec, in_hd, out_rec, out_hd,
			(char **) NULL, (short **) NULL);
		if(add_tag)
			out_rec->tag = dbuf;
		put_fea_rec (out_rec, out_hd, out_strm);
	      }
	    else {

		break;
	    }
	    size = field_size;
	    if(!add_tag) {
	    	d_ptr = (double *) ptr;
	    	f_ptr = (float *) ptr;
	    	l_ptr = (long *) ptr;
	    	s_ptr = (short *) ptr;
	    	c_ptr = ptr;
    	    	dc_ptr = (double_cplx *) ptr;
    	    	fc_ptr = (float_cplx *) ptr;
    	    	lc_ptr = (long_cplx *) ptr;
    	    	sc_ptr = (short_cplx *) ptr;
    	    	bc_ptr = (byte_cplx *) ptr;
	    }
	    else 
		l_ptr =  &(out_rec->tag);
	}
    }

    if (size != field_size) {
	while (size--)
	    switch (type) {
		case DOUBLE: 
		    *d_ptr++ = 0;
		    break;
		case FLOAT: 
		    *f_ptr++ = 0;
		    break;
		case LONG: 
		    *l_ptr++ = 0;
		    break;
		case SHORT: 
		    *s_ptr++ = 0;
		    break;
		case BYTE:
		    *c_ptr++ = 0;
		    break;
		case DOUBLE_CPLX:
		    dc_ptr->real = 0;
		    dc_ptr->imag = 0;
		    dc_ptr++;
		    break;
		case FLOAT_CPLX:
		    fc_ptr->real = 0;
		    fc_ptr->imag = 0;
		    fc_ptr++;
		    break;
		case LONG_CPLX:
		    lc_ptr->real = 0;
		    lc_ptr->imag = 0;
		    lc_ptr++;
		    break;
		case SHORT_CPLX:
		    sc_ptr->real = 0;
		    sc_ptr->imag = 0;
		    sc_ptr++;
		    break;
		case BYTE_CPLX:
		    bc_ptr->real = 0;
		    bc_ptr->imag = 0;
		    bc_ptr++;
		    break;
	    }
	if (new_file && !out_file)
	    put_fea_rec (out_rec, out_hd, out_strm);
	else
	if (get_fea_rec (in_rec, in_hd, in_strm) != EOF) {
	    copy_fea_rec (in_rec, in_hd, out_rec, out_hd,
		    (char **) NULL, (short **) NULL);
	    put_fea_rec (out_rec, out_hd, out_strm);
	}
	    size = field_size;
	    if(!add_tag) {
	    	d_ptr = (double *) ptr;
	    	f_ptr = (float *) ptr;
	    	l_ptr = (long *) ptr;
	    	s_ptr = (short *) ptr;
	    	c_ptr = ptr;
    	    	dc_ptr = (double_cplx *) ptr;
    	    	fc_ptr = (float_cplx *) ptr;
    	    	lc_ptr = (long_cplx *) ptr;
    	    	sc_ptr = (short_cplx *) ptr;
    	    	bc_ptr = (byte_cplx *) ptr;
	    }
	    else 
		l_ptr =  &(out_rec->tag);
    }


    if (out_file || update) {	/* get rest of input records */
	long    extra_count = 0;
	if (type == CHAR)
	    size = 1;
	while (size--)
	    switch (type) {
		case DOUBLE: 
		    *d_ptr++ = 0;
		    break;
		case FLOAT: 
		    *f_ptr++ = 0;
		    break;
		case LONG: 
		    *l_ptr++ = 0;
		    break;
		case SHORT: 
		    *s_ptr++ = 0;
		    break;
		case CHAR: 
		    if (field_size == 1)
			*ptr = ' ';
		    else
		    	strcpy (ptr, " ");
		    break;
		case BYTE:
		    *c_ptr++ = 0;
		    break;
		case DOUBLE_CPLX:
		    dc_ptr->real = 0;
		    dc_ptr->imag = 0;
		    dc_ptr++;
		    break;
		case FLOAT_CPLX:
		    fc_ptr->real = 0;
		    fc_ptr->imag = 0;
		    fc_ptr++;
		    break;
		case LONG_CPLX:
		    lc_ptr->real = 0;
		    lc_ptr->imag = 0;
		    lc_ptr++;
		    break;
		case SHORT_CPLX:
		    sc_ptr->real = 0;
		    sc_ptr->imag = 0;
		    sc_ptr++;
		    break;
		case BYTE_CPLX:
		    bc_ptr->real = 0;
		    bc_ptr->imag = 0;
		    bc_ptr++;
		    break;
	    }
	while (get_fea_rec (in_rec, in_hd, in_strm) != EOF) {
	    copy_fea_rec (in_rec, in_hd, out_rec, out_hd,
		    (char **) NULL, (short **) NULL);
	    put_fea_rec (out_rec, out_hd, out_strm);
	    extra_count++;
	}
	if (extra_count)
	    fprintf (stderr, "addfea: Warning, %ld extra records in %s were copied \naddfea:         to output file after ASCII input exhausted.\n",
		    extra_count, in_file);
    }

/* if updating the original file, we rewind the temp file and then
   copy the data over
*/
    if (update) {

	FILE * t_strm;

	rewind (out_strm);	/* rewind the temp file */
	fclose (in_strm);
	t_strm = out_strm;
	if ((out_strm = fopen (in_file, "w")) == NULL)
	    error2 ("cannot open input file for writing:", in_file, 0);
	else
	    while ((c = getc (t_strm)) != EOF)
		putc (c, out_strm);
    }

    exit(0);

}
